Document Trailing Ops Compatibility (#11685)
microsoft/FluidFramework: packages/dds/README.md

# Distributed Data Structures

Fluid's Distributed Data Structures (DDSes) packages live here.
Documentation for using them can be found [here](../../docs/content/docs/build/dds.md).

This documentation (below) is for people working on DDS implementations.
This document is incomplete and extensions to it are welcome.

## <a name="trailing-ops"></a>Trailing Ops and Compatibility

When initializing a DDS instance from an existing document, first the summary is loaded, and then any ops since that summary are applied.

> These ops are called "Trailing Ops", and they can get persisted as part of the server side summarization when a session ends between summaries.

Author: Yes, that order is better. I have rephrased accordingly.

> Since Fluid supports being used in a mode (and it typically used in this mode) where there is no service that summarizes documents to bake these trailing ops into the DDS level summaries,

Reviewer: ```suggestion
Since Fluid supports being used in a mode (and is typically used in this mode) where there is no service that summarizes documents to bake these trailing ops into the DDS level summaries,
```

> Since Fluid supports being used in a mode (and it typically used in this mode) where there is no service that summarizes documents to bake these trailing ops into the DDS level summaries,

Reviewer 2: Are you looking to draw a distinction here about the mechanism used to store the trailing ops?  If so, maybe call out the possibilities more explicitly, e.g.  "Fluid supports two ways a service might store trailing ops:  either as explicit deltas using DeltaStorage, or baked into the summary itself".

Also -- this document states it is for DDS implementers -- what do they need to know about this nuance?  Seems like this all happens behind the scenes and a DDS implementer doesn't care?

Reviewer: Oh, the trailing ops could be stored *still as independent ops* within a summary? I thought that "baked into the summary" meant "they are already considered when generating the snapshot that is the latest summary", which to me meant that at that point the ops as separate entities did not exist anymore (or at least weren't relevant anymore). If the former is true, then I guess I'm an example that making that a bit more explicit could be good 😄 .

Author: I have rephased this to hopefully be clearer.

I have no idea what DeltaStorage is, so I'm not addressing it here. It seems unnecessary for making the point that we need back compat. If you understand it, and think it would be useful to cover in this document, feel free to add it :).

these trailing ops can live in documents for unbounded amounts of time.

> This means DDSes should generally support processing all past versions of ops that were ever used (in addition to supporting all past summary formats) so that old documents continue to be openable.

Author: I don't know what "delta storage" means in this context, regardless we do require op back compat.

> This means DDSes should generally support processing all past versions of ops that were ever used (in addition to supporting all past summary formats) so that old documents continue to be openable.

Author: I think the new phrasing is now less specific about how the data is stored, so this distinction likley doesn't matter much anymore.

# Distributed Data Structures

Fluid's Distributed Data Structures (DDSes) packages live here.
Documentation for using them can be found [here](../../docs/content/docs/build/dds.md).

This documentation (below) is for people working on DDS implementations.
This document is incomplete and extensions to it are welcome.

## <a name="trailing-ops"></a>Trailing Ops and Compatibility

When a session ends between summaries, the most recent ops might not be included in the latest summary.
These ops are still persisted by the service as "trailing ops".
When a DDS is loaded, first the summary is loaded, then the trailing ops are applied.

> Fluid typically has no service that updates summaries by applying these trailing ops to the DDS.
> Thus these trailing ops may remain in documents indefinitely, only being removed if a client opens, edits and re-summarizes the document.

Reviewer 2: Might also read easier with an order flip.

```suggestion
These trailing ops may remain in documents indefinitely, only being removed if a client opens, edits and re-summarizes the document.  Fluid typically does not provide a service to update summaries by applying these trailing ops to the DDS.
```
